langchain/src/chat_models/universal.ts

```diff
@@ -0,0 +1,827 @@
+ *       })
+ *       .describe("Get the current weather in a given location"),
+ *     name: "GetWeather",
+ *     description: "Get the current weather in a given location",
+ *   }
+ * );
+ *
+ * const getPopulationTool = tool(
+ *   (input) => {
+ *     // Do something with the input
+ *     return JSON.stringify(input);
+ *   },
+ *   {
+ *     schema: z
+ *       .object({
+ *         location: z
+ *           .string()
+ *           .describe("The city and state, e.g. San Francisco, CA"),
+ *       })
+ *       .describe("Get the current population in a given location"),
+ *     name: "GetPopulation",
+ *     description: "Get the current population in a given location",
+ *   }
+ * );
+ *
+ * const configurableModel = await initChatModel("gpt-4", {
+ *   configurableFields: ["model", "modelProvider", "apiKey"],
+ *   temperature: 0,
+ * });
+ *
+ * const configurableModelWithTools = configurableModel.bind({
+ *   tools: [getWeatherTool, getPopulationTool],
+ * });
+ *
+ * const configurableToolResult = await configurableModelWithTools.invoke(
+ *   "Which city is hotter today and which is bigger: LA or NY?",
+ *   {
+ *     configurable: {
+ *       apiKey: process.env.OPENAI_API_KEY,
+ *     },
+ *   }
+ * );
+ *
+ * const configurableToolResult2 = await configurableModelWithTools.invoke(
+ *   "Which city is hotter today and which is bigger: LA or NY?",
+ *   {
+ *     configurable: {
+ *       model: "claude-3-5-sonnet-20240620",
+ *       apiKey: process.env.ANTHROPIC_API_KEY,
+ *     },
+ *   }
+ * );
+ * ```
+ *
+ * @description
+ * This function initializes a ChatModel based on the provided model name and provider.
+ * It supports various model providers and allows for runtime configuration of model parameters.
+ *
+ * Security Note: Setting `configurableFields` to "any" means fields like api_key, base_url, etc.
+ * can be altered at runtime, potentially redirecting model requests to a different service/user.
+ * Make sure that if you're accepting untrusted configurations, you enumerate the
+ * `configurableFields` explicitly.
+ *
+ * The function will attempt to infer the model provider from the model name if not specified.
+ * Certain model name prefixes are associated with specific providers:
+ * - gpt-3... or gpt-4... -> openai
+ * - claude... -> anthropic
+ * - amazon.... -> bedrock
+ * - gemini... -> google-vertexai
+ * - command... -> cohere
+ * - accounts/fireworks... -> fireworks
+ *
+ * @since 0.2.11
+ * @version 0.2.11
+ */
+export async function initChatModel<
+  RunInput extends BaseLanguageModelInput = BaseLanguageModelInput,
+  CallOptions extends ConfigurableChatModelCallOptions = ConfigurableChatModelCallOptions
+>(
+  model?: string,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  fields?: Partial<Record<string, any>> & {
+    modelProvider?: string;
+    configurableFields?: string[] | "any";
+    configPrefix?: string;
+  }
+): Promise<_ConfigurableModel<RunInput, CallOptions>> {
```

Comment:
nit: should this just return something typed as `BaseChatModel`? `_ConfigurableModel` is an implementation detail

Reply:
Remember to update overload signatures too

Reply:
Also should docstring go above all overloads

Reply:
Can't because `bindTools` on `BaseChatModel` is optional, so we'd need to re-implement that `BaseChatModelWithBindTools` class I had before.

Reply:
Hmm ok